Utils/ObjectPool.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace DUCK.Pooling
+{
+	public abstract class PoolableComponent : MonoBehaviour
+	{
+		protected internal virtual void OnAddedToPool()
+		{
+			enabled = false;
+			gameObject.SetActive(false);
+		}
+
+		protected internal virtual void OnRemovedFromPool()
+		{
+			gameObject.SetActive(true);
+			enabled = true;
+		}
+
+		protected internal abstract void CopyFrom(PoolableComponent instance);
+	}
+
+	public static class ObjectPool
+	{
+		private abstract class AbstractPool
+		{
+			public abstract Type ObjectType { get; }
+
+			public abstract void Add(PoolableComponent obj);
+		}
+
+		private class Pool<T> : AbstractPool 
+			where T : PoolableComponent
+		{
+			private Queue<T> pooledObjects = new Queue<T>();
+
+			public override Type ObjectType
+			{
+				get
+				{
+					return typeof(T);
+				}
+			}
+
+			public T Remove()
```

Comment:
Perhaps `Get` would be a better name, as it sounds less destructive, theoretically this is where you would (without pooling) be creating the object, not removing it. Removal is just part of the pooling mechanism.

Reply:
Renamed